deployment/ccip/changeset/cs_chain_contracts.go

```diff
@@ -351,6 +353,283 @@ func UpdateOnRampsDestsChangeset(e deployment.Environment, cfg UpdateOnRampDests
+			}
+			for _, sender := range update.AddedAllowlistedSenders {
+				if sender == (common.Address{}) {
+					return fmt.Errorf("can't allowlist 0-address sender for src=%d, dest=%d", srcSel, destSel)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func UpdateOnRampAllowListChangeset(e deployment.Environment, cfg UpdateOnRampAllowListConfig) (deployment.ChangesetOutput, error) {
+	if err := cfg.Validate(); err != nil {
+		return deployment.ChangesetOutput{}, err
+	}
+	onchain, err := LoadOnchainState(e)
+	if err != nil {
+		return deployment.ChangesetOutput{}, err
+	}
+	var batches []timelock.BatchChainOperation
+	timelocks := make(map[uint64]common.Address)
+	proposers := make(map[uint64]*gethwrappers.ManyChainMultiSig)
+	ctx := e.GetContext()
+	for srcSel, updates := range cfg.UpdatesByChain {
+		txOps := e.Chains[srcSel].DeployerKey
+		//nolint:fatcontext // context is stored outside the loop
+		txOps.Context = ctx
+		if cfg.MCMS != nil {
+			txOps = deployment.SimTransactOpts()
+		}
+		onRamp := onchain.Chains[srcSel].OnRamp
+		args := make([]onramp.OnRampAllowlistConfigArgs, len(updates))
+		for destSel, update := range updates {
+			args = append(args, onramp.OnRampAllowlistConfigArgs{
+				DestChainSelector:         destSel,
+				AllowlistEnabled:          update.AllowListEnabled,
+				AddedAllowlistedSenders:   update.AddedAllowlistedSenders,
+				RemovedAllowlistedSenders: update.RemovedAllowlistedSenders,
+			})
+		}
+		tx, err := onRamp.ApplyAllowlistUpdates(txOps, args)
+		if err != nil {
+			return deployment.ChangesetOutput{}, err
+		}
+		if cfg.MCMS == nil {
+			if _, err := deployment.ConfirmIfNoError(e.Chains[srcSel], tx, err); err != nil {
+				return deployment.ChangesetOutput{}, err
+			}
+		} else {
+			batches = append(batches, timelock.BatchChainOperation{
+				ChainIdentifier: mcms.ChainIdentifier(srcSel),
+				Batch: []mcms.Operation{
+					{
+						To:    onRamp.Address(),
+						Data:  tx.Data(),
+						Value: big.NewInt(0),
+					},
+				},
+			})
+			timelocks[srcSel] = onchain.Chains[srcSel].Timelock.Address()
+			proposers[srcSel] = onchain.Chains[srcSel].ProposerMcm
+		}
+	}
+	if cfg.MCMS == nil {
+		return deployment.ChangesetOutput{}, nil
+	}
+	proposal, err := proposalutils.BuildProposalFromBatches(
+		timelocks,
+		proposers,
+		batches,
+		"update onramp allowlist",
+		cfg.MCMS.MinDelay,
+	)
+	if err != nil {
+		return deployment.ChangesetOutput{}, err
+	}
+	return deployment.ChangesetOutput{
+		Proposals: []timelock.MCMSWithTimelockProposal{*proposal},
+	}, nil
+}
+
+type WithdrawOnRampFeeTokensConfig struct {
+	FeeTokensByChain map[uint64][]common.Address
+	MCMS             *MCMSConfig
+}
+
+func (cfg WithdrawOnRampFeeTokensConfig) Validate(onchain CCIPOnChainState) error {
+	for chainSel, feeTokens := range cfg.FeeTokensByChain {
+		feeQuoter := onchain.Chains[chainSel].FeeQuoter
+		onchainFeeTokens, err := feeQuoter.GetFeeTokens(nil)
```

Comment:
We can pass the environment context through and assign `feeQuoter.GetFeeTokens(&bind.TransactOpts{Context: ctx})`. Applies for all read calls.

Reply:
not need for `&bind.TransactOpts` for read calls though